src/if_statements.py

```diff
@@ -600,27 +648,56 @@ def emit_return(context: Context, node: ReturnNode, body: Body) -> None:
 
 def build_switch_between(
     context: Context,
-    start: SwitchNode,
+    switch: SwitchNode,
     end: Node,
 ) -> SwitchStatement:
-    """Output the subgraph between `start` and `end`, including the jump
-    in the SwitchStatement `start`, but not including `end`"""
-    body = Body(print_node_comment=context.options.debug)
-    jump = switch_jump(context, start)
-    emitted_cases: Set[Node] = set()
-
-    for case in start.cases:
-        if case in context.emitted_nodes:
-            continue
-        if case == end:
-            # We are not responsible for emitting `end`
-            continue
+    """
+    Output the subgraph between `switch` and `end`, but not including `end`.
+    The returned SwitchStatement starts with the jump to the switch's value.
+    """
+    # Try to detect the default node by looking at the switch's parent
+    default_node: Optional[Node] = None
+    if (
+        len(switch.parents) == 1
+        and isinstance(switch.parents[0], ConditionalNode)
+        and switch.parents[0].fallthrough_edge is switch
+        and switch.parents[0].conditional_edge is not end
+    ):
+        default_node = switch.parents[0].conditional_edge
+
+    add_labels_for_switch(context, switch, default_node)
+    switch_index = context.switch_nodes[switch]
+    comment = f" // switch {switch_index}" if switch_index else ""
 
-        body.extend(build_flowgraph_between(context, case, end))
-        # TODO: This could be a `break;` statement
-        if not body.ends_in_jump():
-            emit_goto_or_early_return(context, end, body)
-    return SwitchStatement(jump, body)
+    jump: Optional[Union[Expression, SimpleStatement]]
+    jump = get_block_info(switch).switch_value
+    if not isinstance(jump, SwitchControl):
+        jump = switch_jump(context, switch)
+
+    switch_body = Body(print_node_comment=context.options.debug)
+
+    # Order case blocks by their position in the asm, not by their order in the jump table
+    sorted_cases = sorted(set(switch.cases), key=lambda node: node.block.index)
```

Comment:
are all the `.index`es unique, or can this give non-deterministic results? (oh, why does python make map iteration order deterministic but not set? :()